x-pack/plugin/apm-integration/src/internalClusterTest/java/org/elasticsearch/xpack/apm/ApmIT.java

```diff
@@ -68,4 +81,82 @@ public void testModule() {
         }
         assertTrue(found);
     }
+
+    public void testSearch() throws Exception {
+
+        internalCluster().ensureAtLeastNumDataNodes(2);
+        final int nodeCount = internalCluster().numDataNodes();
+
+        assertAcked(
+            client().admin()
+                .indices()
+                .prepareCreate("test-matching")
+                .setMapping("{\"properties\":{\"message\":{\"type\":\"text\"},\"@timestamp\":{\"type\":\"date\"}}}")
+                .setSettings(
+                    Settings.builder()
+                        .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 1)
+                        .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, nodeCount * 6)
+                )
+        );
+
+        assertAcked(
+            client().admin()
+                .indices()
+                .prepareCreate("test-notmatching")
+                .setMapping("{\"properties\":{\"message\":{\"type\":\"text\"},\"@timestamp\":{\"type\":\"date\"}}}")
+                .setSettings(
+                    Settings.builder()
+                        .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 1)
+                        .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, nodeCount * 6)
+                )
+        );
+
+        ensureGreen("test-matching", "test-notmatching");
+
+        final String matchingDate = "2021-11-17";
+        final String nonMatchingDate = "2021-01-01";
+
+        final BulkRequestBuilder bulkRequestBuilder = client().prepareBulk().setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE);
+
+        for (int i = 0; i < 1000; i++) {
+            final boolean isMatching = randomBoolean();
+            final IndexRequestBuilder indexRequestBuilder = client().prepareIndex(isMatching ? "test-matching" : "test-notmatching");
+            indexRequestBuilder.setSource(
+                "{\"@timestamp\":\"" + (isMatching ? matchingDate : nonMatchingDate) + "\",\"message\":\"\"}",
+                XContentType.JSON
+            );
+            bulkRequestBuilder.add(indexRequestBuilder);
+        }
+
+        assertFalse(bulkRequestBuilder.execute().actionGet(10, TimeUnit.SECONDS).hasFailures());
+
+        APMTracer.CAPTURING_SPAN_EXPORTER.clear();
+
+        client().prepareSearch()
+            .setQuery(new RangeQueryBuilder("@timestamp").gt("2021-11-01"))
+            .setSearchType(SearchType.QUERY_THEN_FETCH)
+            .setPreFilterShardSize(1)
+            .execute()
+            .actionGet(10, TimeUnit.SECONDS);
+
+        assertBusy(() -> {
```

Comment:
Should we have all the spans captured by the time the search is complete?

Reply:
Ah yes you're right, we're only using the `BatchSpanProcessor` for pushing them over the wire. (I extracted this PR from a branch in which I'm using the `BatchSpanProcessor` for everything).